cpp/src/utilities/wrapper_types.hpp

```diff
@@ -77,11 +77,19 @@ struct wrapper
   using value_type = T;                                          ///< The underlying fundamental type of the wrapper
   value_type value;                                              ///< The wrapped value
 
+  template <gdf_dtype dtype = type_id,
+            std::enable_if_t<(dtype != GDF_BOOL8)> * = nullptr>
   CUDA_HOST_DEVICE_CALLABLE
   constexpr explicit wrapper(T v) : value{v} {}
 
-  CUDA_HOST_DEVICE_CALLABLE
-  explicit operator value_type() const { return this->value; }
+  template <gdf_dtype dtype = type_id,
+            std::enable_if_t<(dtype == GDF_BOOL8)> * = nullptr>
+  CUDA_HOST_DEVICE_CALLABLE constexpr explicit wrapper(T v)
```

Comment:
Why does it still need to be `explicit`? We should be able to do 
```c++
int a = 1;
float b = 1.0;
cudf::bool8 b = (a && b);
```
I understand for other wrapper types. Or is it required so that `type_dispatcher` works correctly.

Reply:
If it wasn't `explicit`, then something like this could happen:

```
void foo(cudf::bool8 b);
...
char c;
foo(c); // This will work if `wrapper::wrapper(char)` isn't `explicit`, which probably isn't what you want. 
```

Reply:
I don't believe removing the `explicit` from the constructor would solve your above example anyways. 

To do what you suggest, we could add a `wrapper::wrapper(bool)` constructor that is only defined for `cudf::bool8`.

Reply:
Just for fun I tried removing the `explicit` keyword from the `bool8` constructor, and got a very perplexing compiler error:
```
../src/utilities/wrapper_types.hpp: In substitution of ‘template<gdf_dtype dtype, std::enable_if_t<(dtype == GDF_BOOL8), void>* <anonymous> > constexpr cudf::detail::wrapper<T, type_id>::wrapper(T) [with gdf_dtype dtype = (gdf_dtype)7u; std::enable_if_t<(dtype == GDF_BOOL8), void>* <anonymous> = <missing>]’:
../src/utilities/wrapper_types.hpp:86:56:   recursively required by substitution of ‘template<gdf_dtype dtype, std::enable_if_t<(dtype == GDF_BOOL8), void>* <anonymous> > constexpr cudf::detail::wrapper<T, type_id>::wrapper(T) [with gdf_dtype dtype = (gdf_dtype)7u; std::enable_if_t<(dtype == GDF_BOOL8), void>* <anonymous> = <missing>]’
../src/utilities/wrapper_types.hpp:86:56:   required by substitution of ‘template<gdf_dtype dtype, std::enable_if_t<(dtype == GDF_BOOL8), void>* <anonymous> > constexpr cudf::detail::wrapper<T, type_id>::wrapper(T) [with gdf_dtype dtype = (gdf_dtype)7u; std::enable_if_t<(dtype == GDF_BOOL8), void>* <anonymous> = <missing>]’
../src/utilities/wrapper_types.hpp:97:61:   required by substitution of ‘template<gdf_dtype the_type, class T_out, typename std::enable_if<((the_type == GDF_BOOL8) && std::is_arithmetic<T_out>::value), int>::type* <anonymous> > cudf::detail::wrapper<T, type_id>::operator T_out() const [with gdf_dtype the_type = (gdf_dtype)7u; T_out = bool; typename std::enable_if<((the_type == GDF_BOOL8) && std::is_arithmetic<T_out>::value), int>::type* <anonymous> = <missing>]’
../src/utilities/wrapper_types.hpp:379:31:   required from here
../src/utilities/wrapper_types.hpp:86:56: fatal error: template instantiation depth exceeds maximum of 900 (use -ftemplate-depth= to increase the maximum)
             std::enable_if_t<(dtype == GDF_BOOL8)> * = nullptr>
```
Not sure what the root cause of that is.

Reply:
I've seen this one too when I was trying to remove the `explicit` keyword (just for bool8) for some issue I was facing in writing logical ops. That's why I brought this up. Thought you might know what's this compiler error.